pkg/ingestor/parser/common/scanner/scanner.go

```diff
@@ -0,0 +1,119 @@
+//
+// Copyright 2024 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scanner
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	osv_scanner "github.com/google/osv-scanner/pkg/osv"
+
+	"github.com/guacsec/guac/pkg/assembler"
+	"github.com/guacsec/guac/pkg/certifier/components/root_package"
+	"github.com/guacsec/guac/pkg/certifier/osv"
+	"github.com/guacsec/guac/pkg/events"
+	"github.com/guacsec/guac/pkg/handler/processor"
+	"github.com/guacsec/guac/pkg/ingestor/parser/common"
+	"github.com/guacsec/guac/pkg/ingestor/parser/vuln"
+	"github.com/guacsec/guac/pkg/version"
+)
+
+// PurlsToScan takes a slice of purls and bulk queries OSV (skipping purls that start with "pkg:guac").
+// Once the query returns, an attestation is generated and passed to the vulnerability parser for ingestion
+func PurlsToScan(ctx context.Context, purls []string) ([]assembler.VulnEqualIngest, []assembler.CertifyVulnIngest, error) {
+	type docResult struct {
+		doc    *processor.Document
+		docErr error
+	}
+
+	docChan := make(chan docResult, 1)
+
+	go func(ctx context.Context, purls []string, docChan chan<- docResult) {
```

Comment:
Is there a good reason to do this in a Go routine? Do we believe that doing this:
```go
		vulnParser.Parse(ctx, response.doc)
		preds := vulnParser.GetPredicates(ctx)
		common.AddMetadata(preds, nil, response.doc.SourceInformation)
```

At the same time in a separate thread will be a significant performance improvement, or do we think that the GC will collect `docResult` so that we don't use too much memory?

Reply:
Yeah, I was going back and forth between using the go routine or not. Let me check the performance and make the necessary changes.

Reply:
so time-wise with go routine:
```
go run ./cmd/guacone collect files --query-vuln=true ../guac-data/docs  9.14s user 1.90s system 13% cpu 1:19.07 total
```

Without go routine:
```
go run ./cmd/guacone collect files --query-vuln=true ../guac-data/docs  9.09s user 1.74s system 13% cpu 1:22.05 total
```

So a couple of seconds improvement overall...so I can keep or remove?

Reply:
I can remove or keep the go routine....

Reply:
Yeah, seems fine. I know we use the same pattern elsewhere as well, so there is consistency.